python/ray/train/v2/_internal/execution/checkpoint/report_handler.py

```diff
@@ -0,0 +1,112 @@
+from collections import deque
+from typing import TYPE_CHECKING, Deque, List, Optional
+
+from ray.train.v2._internal.execution.callback import (
+    ReportCallback,
+    WorkerGroupCallback,
+)
+from ray.train.v2._internal.execution.worker_group import WorkerGroup, WorkerGroupStatus
+
+if TYPE_CHECKING:
+    from ray.train._internal.session import _TrainingResult
+
+
+class ReportHandler(WorkerGroupCallback):
+    """Handle `ray.train.report` calls from multiple workers.
+
+    Execute hooks once all workers have reported a training result.
+    """
+
+    def __init__(self, report_callbacks: List[ReportCallback]):
+        # Number of workers in the current worker group. It is initialized
+        # to be None. It is set to the number of workers when it receives the
+        # worker group status for the first time.
+        # When a worker group shutdown, self._num_workers is set to None,
+        # waiting to be updated when a new worker group status is received again.
+        self._num_workers: Optional[int] = None
+        # A list of queues holding training results from workers.
+        self._training_result_queues: Optional[List[Deque[_TrainingResult]]] = None
+
+        self._report_callbacks = report_callbacks
+
+    # --------------------------
+    # WorkerGroupCallback
+    # --------------------------
+
+    def after_worker_group_poll_status(
+        self, worker_group_status: WorkerGroupStatus
+    ) -> None:
```

Comment:
Note: we shouldn't have too many callbacks implementing this method since it might accumulate a lot of overhead every health check period (ex: 2 seconds).

`UserCallbackInvoker` also implements this now to capture the worker group failures.